### PR TITLE
feat(linter): eslint/no-iterator

### PR DESCRIPTION
**Pull request by JoSeBu1**
Rule detail: https://eslint.org/docs/latest/rules/no-iterator

### CODE REVIEW

**Comment by overlookmotel** [comment]
@JoSeBu1 In case you're wondering why I have rebased this on main, I made a mistake when sharding benchmarks, which lead to benchmarks results not being submitted correctly on PRs from forks. Now fixed I hope (#2759) so am making sure it now works by re-running benchmarks on PRs.

**Comment by overlookmotel** [comment]
Yup, it works! Am going to rebase all other PRs submitted today so they get benchmarked on CodSpeed too.

**Comment by JoSeBu1** [comment]
@overlookmotel Before all, your work is amazing, I'm too noob with Rust but I updated my PRs that are open

**Comment by overlookmotel** [comment]
> I updated my PRs that are open

Thanks for doing that. And thanks for the compliment!

**Comment by JoSeBu1** [comment]
No worries, I just want to support this extraordinary project by dedicating a little of my time, and the little knowledge I have with Rust (I started a couple of days ago). I think I can wait as long as I have to hehe

File: crates/oxc_linter/src/rules/eslint/no_iterator.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    /// The __iterator__ property was a SpiderMonkey extension to JavaScript that could be used to create custom iterators that are compatible with JavaScript’s for in and for each constructs. However, this property is now obsolete, so it should not be used. Here’s an example of how this used to work:
+    ///
+    /// ### Example
+    /// ```javascript
+    /// Foo.prototype.__iterator__ = function() {
+    ///     return new FooIterator(this);
+    /// }
+    /// ```
+    NoIterator,
+    style
```
[the remark this answers is not in the document]
Comment (JoSeBu1) [reply crates/oxc_linter/src/rules/eslint/no_iterator.rs:33]: Or maybe restriction?

[the remark this answers is not in the document]
Comment (mysteryven) [reply crates/oxc_linter/src/rules/eslint/no_iterator.rs:33]: Let's change it to `restriction`.

[the remark this answers is not in the document]
Comment (JoSeBu1) [reply crates/oxc_linter/src/rules/eslint/no_iterator.rs]: done

**Review by mysteryven — APPROVED** [review]
Thank you!

**Pull request merged**